src/hotspot/share/opto/parseHelper.cpp

```diff
@@ -869,9 +865,248 @@ Node* PEAState::get_java_oop(ObjID id) const {
   return obj;
 }
 
-AllocationStateMerger::AllocationStateMerger(PEAState& target) : _state(target) {}
+static SafePointNode* attached_sfpt(RegionNode* shim) {
+  SafePointNode* sfpt = nullptr;
+  for (DUIterator_Fast imax, i = shim->fast_outs(imax); i < imax; i++) {
+    Node* node = shim->fast_out(i);
+    if (node->is_SafePoint() && node->outcnt() == 0) {
+        assert(sfpt == nullptr || sfpt == node, "sanity");
+        sfpt = node->as_SafePoint();
+    }
+  }
+  return sfpt;
+}
+
+static bool matched(SafePointNode* sfpt, ObjID id, VirtualState* vs) {
+  PEAState& as = sfpt->jvms()->alloc_state();
+  return as.contains(id) && as.equals(id, vs);
+}
 
-void AllocationStateMerger::merge(PEAState& newin, GraphKit* kit, RegionNode* region, int pnum) {
+// update the internal phi node in virtual objects.
```

Comment:
Can we have a better name for `color`. I have no idea what it means.

Can we describe what needs to be updated?

Reply:
for DFS, we need to prevent from revisiting a node which has been visited. 
the standard DFS algorithm using coloring scheme: white (unvisited), gray(processing) and black (processed). 

this is simplified version. we don't need to distinct gray and black. we just mark it visited.